NextcloudTalk/NCChatViewController.m

```diff
@@ -1033,6 +1037,47 @@ - (void)setFailedStatusToMessageWithReferenceId:(NSString *)referenceId
     [self.tableView endUpdates];
 }
 
+#pragma mark - Message expiration
+
+- (void)startObservingExpiredMessages
+{
+    _messageExpirationTimer = [NSTimer scheduledTimerWithTimeInterval:30.0 target:self selector:@selector(removeExpiredMessages) userInfo:nil repeats:YES];
+}
+
+- (void)removeExpiredMessages
+{
+    dispatch_async(dispatch_get_main_queue(), ^{
+        NSInteger currentTimestamp = [[NSDate date] timeIntervalSince1970];
+        for (NSInteger i = 0; i < self->_dateSections.count; i++) {
+            NSDate *keyDate = [self->_dateSections objectAtIndex:i];
+            NSMutableArray *messages = [self->_messages objectForKey:keyDate];
+            NSMutableArray *deleteMessages = [NSMutableArray new];
+            for (NSInteger j = 0; j < messages.count; j++) {
+                NCChatMessage *currentMessage = messages[j];
+                NSInteger messageExpirationTime = currentMessage.expirationTimestamp;
+                if (messageExpirationTime > 0 && messageExpirationTime <= currentTimestamp) {
+                    [deleteMessages addObject:currentMessage];
+                }
+            }
+            if (deleteMessages.count > 0) {
+                [messages removeObjectsInArray:deleteMessages];
+                if (messages.count > 0) {
+                    [self.tableView beginUpdates];
```

Comment:
Move `beginUpdate`/`endUpdates` out of the if-clause as we need it in both cases?

Reply:
I used to modify the data source first and then call `beginUpdate` / `endUpdates`.
But it seems that modifying the data source inside `beginUpdate` / `endUpdates` block is also possible ;)